### PR TITLE
feat(anomaly): Raft cluster (Cluster V2) health detection + gossip-mode gating

### PR DESCRIPTION
**Pull request by KIvanow**
## Summary

Adds first-class observability for **Valkey's Raft-based cluster bus (Cluster V2, `cluster-protocol raft`)** — the consensus rework on the upstream [`cluster-v2`](https://github.com/valkey-io/valkey/tree/cluster-v2) branch that replaces the gossip cluster bus (and fixes, at the source, the class of bugs behind our #2090 / #2261 detectors).

**The observable surface and failure behaviours were verified against a live 3-node Raft cluster** built from the `cluster-v2` branch (not guessed from diffs). In Raft mode `CLUSTER INFO` gains: `cluster_raft_role` (leader/follower/candidate/pre-candidate/joiner), `cluster_raft_current_term`, `cluster_raft_commit_index`, `cluster_raft_last_applied`, `cluster_raft_log_entries`, `cluster_raft_leader`.

## Changes
- **`raft-health-detector.ts`** — `parseRaftState()` (returns `null` in gossip mode, which is the mode gate) + `isRaftLeaderless()`.
- **`anomaly.service.ts` → `detectRaftHealth()`** on a dedicated `MetricType.RAFT_HEALTH`:
  - **CRITICAL — leaderless / quorum loss:** `cluster_state:fail` while this node isn't the leader. Under Raft that means a majority is unreachable → the commit log freezes and writes are refused (`CLUSTERDOWN`). Gated on persistence so a *healthy* failover (which keeps `cluster_state:ok` and bumps the term once) never trips it.
  - **WARNING — election churn:** the term advancing repeatedly within a window. Because of the **pre-vote** protocol the term only rises on a *completed* election, so rapid term growth is genuine flapping leadership. (Observed live: quorum loss does **not** inflate the term — so churn ≠ quorum loss.)
- **Mode gate:** the gossip-era topology detectors (duplicate-primary #2261, stuck-replica #2090) are **skipped in Raft mode**, where topology is consensus-managed; `detectRaftHealth` covers Raft instead.

## Verification
- Field names + failover / quorum-loss behaviours captured from a **live `cluster-protocol raft` cluster** (docker build of `cluster-v2`); those exact values are used as test fixtures.
- Tests: pure detector (`parseRaftState`/`isRaftLeaderless`) + service tests (leaderless persistence gate, election churn, healthy-failover suppression, and gossip-vs-Raft detector routing).
- Note: the full anomaly suite (187 tests incl. these) passed on a pre-#319 base; the local run on top of current master is blocked only by an unrelated dependency-staleness artifact (#319's new `@opentelemetry/*` deps not yet installed locally) — **CI runs the authoritative full suite**.

## Follow-ups (cross-node signals needing fan-out)
- Commit-lag / stuck-follower (compare `cluster_raft_commit_index` across nodes).
- Leader-disagreement / >1 leader across nodes.
- "Cluster V2 health" dashboard + version/mode surfacing.

## Checklist
- [x] Unit / integration tests added
- [ ] Docs added / updated
- [ ] [Roborev](https://www.roborev.io/) review passed — run `roborev review --branch` or `/roborev-review-branch` in Claude Code *(internal)*
- [ ] Competitive analysis done / discussed *(internal)*
- [ ] Blog post about it discussed *(internal)*

---

> [!NOTE]
> **Medium Risk**
> Changes cluster incident detection and routing (Raft vs gossip) plus operator-facing quorum alerts; logic is heavily tested but misclassification could affect on-call response during outages.
> 
> **Overview**
> Adds **Valkey Cluster V2 (Raft)** observability end-to-end: backend `raft_health` anomalies and web surfaces that stay accurate when `CLUSTER INFO` alone is misleading.
> 
> **Backend:** New `raft-health-detector` and `detectRaftHealth()` on `MetricType.RAFT_HEALTH`. **CRITICAL** quorum loss is detected from sustained leader-seeking with a frozen commit index (not `cluster_state:fail`), with windows scaled from `cluster-node-timeout`, storage-backed pin/re-emit after dismiss, and auto-resolve on recovery. **WARNING** for rapid term advances (election churn). Raft mode skips gossip duplicate-primary/stuck-replica detectors and is remembered across transient `CLUSTER INFO` failures. Prometheus/OTLP emits in `addAnomaly` are isolated so one failure cannot skip quorum detection.
> 
> **Frontend:** `RaftHealthBanner` on the anomaly dashboard (date-range–independent active feed) with a remediation runbook and server-confirmed dismiss. `RaftHealthPanel` on the cluster overview parses `cluster_raft_*` fields, distinguishes slots vs electing vs **no quorum** via `outageActive` from active critical events, and hides in gossip mode.
> 
> Extensive unit tests cover detector logic, service edge cases, and UI behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 945f725edc16a3ba245b9ecf030d76f62fc1df0d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by KIvanow** [comment]
### Correction found by running it against a live Raft cluster

Stood up a real 3-node Cluster V2 (Raft) group and killed the majority. The original leaderless signal (`cluster_state:fail && role != leader`) **never fired** — the surviving replica keeps reporting `cluster_state:ok` (its slots stay "covered"), `CLUSTER NODES` shows every peer `connected`, and the term does **not** inflate (pre-vote). 60s of real quorum loss, `cluster_state` stayed `ok` the whole time.

**Fix (this push):** replace `isRaftLeaderless()` with `isRaftSeeking()` and run a temporal *leaderless watch* — opens when the node starts seeking (`candidate`/`pre-candidate`), persists across the `follower`↔`pre-candidate` flap, and fires CRITICAL only if no leader emerges and the commit index makes no progress for `RAFT_LEADERLESS_MIN_MS`. Healthy failover closes the watch first (a new leader commits a no-op within ~1s → commit index advances); an idle follower never seeks, so a merely-frozen commit index can't false-positive.

**Verified end-to-end** (enterprise license, live cluster): healthy → no event; kill majority → CRITICAL fires at the 10s gate with `role=pre-candidate, term=3, commit index frozen at 17, cluster_state still "ok"`; recover → watch closes. 120/120 anomaly-detection tests pass, incl. new regression cases (cluster_state:ok outage, oscillation, healthy-failover suppression, commit-progress recovery, idle-follower non-alert).

The companion panel change is in #327.

**Review by jamby77 — APPROVED** [review]
Approving. The last batch closes out everything from the earlier rounds: the resolve-retry now works on both leader- and follower-path recoveries, the re-emit/dedupe decision rides the same authoritative storage-backed feed the UI reads, best-effort emits are isolated so a telemetry throw can't abort the rest of the poll, and the node-timeout cache got a proper backoff/recheck. I traced every commented invariant against the code and found no contradictions, and the test suite pins each prior finding with a named regression test. Nice work.

One cleanup to do (fine as a fast-follow commit):

**Stale retry-on-failure comments.** The churn-path comments still describe semantics that no longer exist: "an addAnomaly failure doesn't drop the election ... the WARNING is retried on the next poll" and `// dedupe: clear only after a successful emit`. Since the last commit wrapped the Prometheus/OTLP emits in try/catch (storage was already caught), `addAnomaly` can no longer reject — so there is no failure path, the "retry" is dead text, and the clear is unconditional in practice. Please reword both so the next reader doesn't design against a guarantee that isn't there.

https://github.com/BetterDB-inc/monitor/blob/355217c32b69fc2ee19ddee010c6fe9da6e7c669/proprietary/anomaly-detection/anomaly.service.ts#L1046-L1050

https://github.com/BetterDB-inc/monitor/blob/355217c32b69fc2ee19ddee010c6fe9da6e7c669/proprietary/anomaly-detection/anomaly.service.ts#L1079-L1081

Three narrow edge cases, for the record as follow-ups — none blocking, each needs a compound coincidence and self-heals or is manually recoverable:

1. **Restart/removal mid-outage orphans the active incident.** The recovery-time resolve is gated solely on the in-memory `raftLeaderlessActive` flag, which `onConnectionRemoved` deletes and a service restart clears. If either happens while a CRITICAL is active in storage, recovery never resolves it — it stays pinned until manually dismissed.

https://github.com/BetterDB-inc/monitor/blob/355217c32b69fc2ee19ddee010c6fe9da6e7c669/proprietary/anomaly-detection/anomaly.service.ts#L1169-L1176

2. **Dismissal during a quiet beat of the oscillation.** The fire/re-emit block is gated on `&& seeking`, so if an operator dismisses the incident while the role is momentarily `follower`, nothing re-emits until the node next seeks — up to ~2x node-timeout of green UI during a live outage. I get why the guard exists (the mismatched fire/recovery clock false-CRITICAL), so this is a genuine tension, not an oversight; worth a test that combines the oscillation with a mid-outage dismiss if you ever revisit.

https://github.com/BetterDB-inc/monitor/blob/355217c32b69fc2ee19ddee010c6fe9da6e7c669/proprietary/anomaly-detection/anomaly.service.ts#L1126-L1128

3. **Duplicate emit under eviction + storage outage.** The storage-down fallback comment assumes "an already-emitted one is still in the cache", but the ring is capped at 1000 across all connections — if the live event has been evicted and storage is down at the fire-check, the dedupe sees nothing and emits a second CRITICAL for the same outage.

https://github.com/BetterDB-inc/monitor/blob/355217c32b69fc2ee19ddee010c6fe9da6e7c669/proprietary/anomaly-detection/anomaly.service.ts#L986-L993

File: proprietary/anomaly-detection/anomaly.service.ts [hunk 1]
```diff
@@ -817,6 +869,317 @@ export class AnomalyService extends MultiConnectionPoller implements OnModuleIni
...
+      // The FIRST fire is gated on the node actively seeking: the fire clock runs
+      // from watch-open (`since`) but recovery (`settled`) runs from the last seek,
+      // so a node that sought for a while then quietly recovered into an idle
+      // follower must not trip a CRITICAL in the gap before `settled` closes the
+      // watch. Once the outage is confirmed (raftLeaderlessActive set), we keep the
+      // pin alive on EVERY poll regardless of the follower↔pre-candidate flap — so
+      // a dismiss landing on a follower beat is re-pinned immediately, not left
+      // green until the next seek.
+      const outageConfirmed = this.raftLeaderlessActive.get(ctx.connectionId) === true;
+      if (timestamp - since >= fireMs && (seeking || outageConfirmed)) {
```
[the remark this answers is not in the document]
Comment (KIvanow) [reply proprietary/anomaly-detection/anomaly.service.ts:1132]: accepted adge case
